### PR TITLE
fix(input): support both endContent and isClearable at the same time

### LINKED ISSUE #2254: [BUG] - Input props endContent and isClearable cannot be used at the same time
### NextUI Version

2.2.9

### Describe the bug

Input cannot use endContent and isClearable together

### Your Example Website or App

_No response_

### Steps to Reproduce the Bug or Issue

<Input
        placeholder="Search Tick"
        variant="bordered"
        endContent={<Search size={16} />}
        isClearable
      />

### Expected behavior

1: show endContent
2: isClearable work

### Screenshots or Videos

_No response_

### Operating System Version

macOS

### Browser

Chrome

### PR DESCRIPTION
**Pull request by kght6123**
- closes: #2254
- closes: #2145

## 📝 Description

Improved UI when both endContent and isClearable are specified.

## ⛳️ Current behavior (updates)

endContent replaces the Clear button.

## 🚀 New behavior

Both endContent and Clear buttons are visible.

## 💣 Is this a breaking change (Yes/No):

No.

Currently, both endContent and isClearable do not work as desired.
It's confusing that endContent is replaced by a Clear button, and I don't think there's any use for it.

## 📝 Additional Information

スクリーンショット 2024-02-02 0 04 17

スクリーンショット 2024-02-02 0 04 45

## Summary by CodeRabbit

- **New Features**
	- Enhanced input component to display both Clear button and end content simultaneously for improved user experience.
	- Introduced dynamic styling for the clear button based on the width of end content.
	- Added a new story for the input component showcasing clearable functionality with visual enhancements.

- **Bug Fixes**
	- Improved responsiveness and visual consistency of the input component, particularly when clearable.

### CODE REVIEW

**Comment by jrgarciadev** [comment]
Hi @kght6123, thank you for submitting this pull request (PR). I will review it shortly

File: packages/core/theme/src/components/input.ts [hunk 1]
```diff
@@ -58,6 +58,8 @@ const input = tv({
       "cursor-pointer",
       "active:!opacity-70",
       "rounded-full",
+      // exist end content
+      "data-[has-end-content=true]:right-9",
```
Comment (jrgarciadev) [inline packages/core/theme/src/components/input.ts:62]: Hi @kght6123,

Thank you for submitting the Pull Request! 🙏🏻 To ensure the `endContent` is always visible regardless of its width, I suggest we make an adjustment. We can utilize the [getBoundingClientRect](https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect) method on the endContent element, if it exists. By doing so, we can dynamically calculate and apply an appropriate margin-right to the clear button, factoring in the actual width of the endContent. This approach will help maintain the visibility and alignment of elements within the layout.

Comment (kght6123) [reply packages/core/theme/src/components/input.ts:62]: @jrgarciadev 
Thank you for the great advice. I believe that to call `getBoundingClientRect`, one must reference `endContent` with React's ref and access the HTML Element. Do you have any good suggestions for this?
 I can't think of a good idea on my own, so please let me know.

Comment (jrgarciadev) [reply packages/core/theme/src/components/input.ts:62]: Hi @kght6123 what about adding a wrapper to the `endContent`?

```js

  const [endContentWidth, setEndContentWidth] = useState(0);

  const endContentWrapperRef = useRef(null)

  useEffect(() => {
    if (endContentRef.current) {
      setWidth(endContentRef.current.getBoundingClientRect().width);
    }
  }, [endContent]);

...
if (isClearable) {
      return (
        <div ref={endContentWrapperRef}>
          <span {...getClearButtonProps()}>{<CloseFilledIcon />}</span>
          {endContent}
        </div>
      );
    }
  ```

**Review by jrgarciadev — CHANGES_REQUESTED** [review]

File: packages/core/theme/src/components/input.ts (hunk 1 above)
Comment (kght6123) [reply packages/core/theme/src/components/input.ts:62]: thank you!

I'll try to fix it with your suggestion.

Comment (kght6123) [reply packages/core/theme/src/components/input.ts]: @jrgarciadev
We have corrected the content you pointed out, so please check again.

File: .changeset/eight-moose-dance.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+---
+"@nextui-org/input": major
+"@nextui-org/theme": major
```
Comment (wingkwong) [inline .changeset/eight-moose-dance.md:3]: change to `patch` instead of `major`

**Review by wingkwong — CHANGES_REQUESTED** [review]
1. please resolve conflicts
2. please handle RTL as well. The clear button should be on the left hand side in RTL.

image

**Comment by wingkwong** [comment]
@kght6123 please resolve conflicts and follow up those comments. thanks.

File: .changeset/eight-moose-dance.md (hunk 2 above)
Comment (jrgarciadev) [inline .changeset/eight-moose-dance.md:3]:
```suggestion
"@nextui-org/input": patch
"@nextui-org/theme": patch
```

**Comment by jrgarciadev** [comment]
Hey @kght6123 please update this PR, handle RTL and fix the conflicts 🙏🏻

**Comment by kght6123** [comment]
@jrgarciadev @wingkwong 
I fixed the conflict and am currently working on it.

I don't have much experience with RTL so I'm learning as I go.

I understand that I need to support bidirectional text and that the clear button should appear on the left side when right-to-left (RTL).

Currently it's on the right, so I'd like to fix this so that it appears on the left side when RTL.

**Review by wingkwong — CHANGES_REQUESTED** [review]
please try to follow the select one. (ref: pr3746)

**Comment by kght6123** [comment]
@wingkwong 
In RTL mode, the clear button is now on the other side.
スクリーンショット 2024-11-08 0 52 12

> please try to follow the select one. (ref: pr3746)
Where in PR 3746 should I refer to? Can you give me more details?

**Comment by wingkwong** [comment]
@kght6123 that PR does the same thing but in select component. I would prefer the same approach rather maintaining two sets of implementations.

**Comment by wingkwong** [comment]
@kght6123 any update?

**Comment by cini9** [comment]
@wingkwong I saw that you closed this PR with unmerged commits. Is this being implemented somewhere else ?

**Comment by wingkwong** [comment]
I think not. I think it was closed bc it was pointing release-next branch which was deleted. One could create a PR with the missing RTL implementation.

**Comment by cini9** [comment]
Was anything merged with only the isClearable + endContent functionality ?

**Comment by wingkwong** [comment]
I think not.

**Comment by cini9** [comment]
Is there a plan to fix this for any of the next releases ?

**Comment by wingkwong** [comment]
depending if someone picks up this PR

**Comment by zdychacek** [comment]
omg

**Pull request closed without merging**